### PR TITLE
Reduce list operation calls when pulling segments from S3

### PR DESCRIPTION
**Pull request by a2l007**
While fetching segments from S3, it presently creates an object summary(LIST operation) for the segment before proceeding to GET the object and so the number of LIST ops are proportional to the number of segments. Since LIST ops are more expensive compared to GET, it is desirable to reduce the number of list ops especially if the LIST limit Is much smaller than for GETs.

This PR lazily creates the object summary  since it isn't really required for pulling segments since the bucket and prefix can be retrieved from the URI and the check to validate if the object is present in the bucket is already done before attempting to pull the segment. This reduces the list operations down to zero while pulling segments.

<hr>

This PR has:
- [x] been self-reviewed.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Review by FrankChen021 — APPROVED** [review]
👍

File: extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/S3DataSegmentPuller.java [hunk 1]
```diff
@@ -231,6 +229,8 @@ public Writer openWriter()
       @Override
       public long getLastModified()
       {
+        final S3ObjectSummary objectSummary =
```
Comment (kfaraz) [inline extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/S3DataSegmentPuller.java:232]: Making the LIST call here every time `getLastModified` is called seems costly.
Maybe initialize it lazily but only once the way `s3Object` is being initialized in `openInputStream`?

**Review by kfaraz — CHANGES_REQUESTED** [review]

File: extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/S3DataSegmentPuller.java (hunk 1 above)
Comment (kfaraz) [reply extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/S3DataSegmentPuller.java:232]: You could also check if `s3Object` is already initialized and just use `s3Object.getObjectMetadata().getLastModified()`.
If not initialized, then you could go for the LIST (or GET if it is cheaper).

**Review by kfaraz — APPROVED** [review]

**Comment by a2l007** [comment]
Thanks for the review @FrankChen021 @kfaraz

**Pull request merged**